internal/appsec/waf.go

```diff
@@ -374,43 +363,39 @@ func newGRPCWAFEventListener(handle *wafHandle, addresses map[string]struct{}, t
 			internalRuntimeNs.Add(internal)
 			nbTimeouts.Add(wafCtx.TotalTimeouts())
 
-			if len(event) == 0 {
-				return
+			if wafResult.HasEvents() {
+				log.Debug("appsec: attack detected by the grpc waf")
+				nbEvents.Inc()
+				mu.Lock()
+				defer mu.Unlock()
+				events = append(events, wafResult.Events...)
 			}
-			log.Debug("appsec: attack detected by the grpc waf")
-			nbEvents.Inc()
-			mu.Lock()
-			events = append(events, event)
-			mu.Unlock()
 		}))
 
 		op.On(grpcsec.OnHandlerOperationFinish(func(op *grpcsec.HandlerOperation, _ grpcsec.HandlerOperationRes) {
 			defer wafCtx.Close()
-			rInfo := handle.RulesetInfo()
-			addWAFMonitoringTags(op, rInfo.Version, overallRuntimeNs.Load(), internalRuntimeNs.Load(), nbTimeouts.Load())
+			wafDiags := handle.Diagnostics()
+			addWAFMonitoringTags(op, wafDiags.Version, overallRuntimeNs.Load(), internalRuntimeNs.Load(), nbTimeouts.Load())
 
 			// Log the following metrics once per instantiation of a WAF handle
 			monitorRulesOnce.Do(func() {
-				addRulesMonitoringTags(op, rInfo)
+				addRulesMonitoringTags(op, wafDiags)
 				op.AddTag(ext.ManualKeep, samplernames.AppSec)
 			})
 
-			addSecurityEvents(op, limiter, events...)
+			addSecurityEvents(op, limiter, events)
 		}))
 	})
 }
 
-func runWAF(wafCtx *waf.Context, values map[string]interface{}, timeout time.Duration) ([]byte, []string) {
-	matches, actions, err := wafCtx.Run(values, timeout)
-	if err != nil {
-		if err == waf.ErrTimeout {
-			log.Debug("appsec: waf timeout value of %s reached", timeout)
-		} else {
-			log.Error("appsec: unexpected waf error: %v", err)
-			return nil, nil
-		}
+func runWAF(wafCtx *waf.Context, values waf.RunAddressData, timeout time.Duration) waf.Result {
+	result, err := wafCtx.Run(values, timeout)
+	if err == waf.ErrTimeout {
+		log.Debug("appsec: waf timeout value of %s reached", timeout)
+	} else if err != nil {
+		log.Error("appsec: unexpected waf error: %v", err)
 	}
-	return matches, actions
+	return result
```

Comment:
It'd need to be made into returning 3 different values, so the derived data is also available...

I'd say if we want to make sure actions are dealt with, we should return a composite object that tracks whether it's been used or not... and have a defer check for outstanding actions... Otherwise, this is just good intentions (it's so easy to assign to `_`!)